vncviewer/BaseTouchHandler.cxx

```diff
@@ -0,0 +1,149 @@
+/* Copyright 2019 Aaron Sowry for Cendio AB
+ * Copyright 2019-2020 Pierre Ossman for Cendio AB
+ * 
+ * This is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ * 
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ * 
+ * You should have received a copy of the GNU General Public License
+ * along with this software; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307,
+ * USA.
+ */
+
+#ifdef HAVE_CONFIG_H
+#include <config.h>
+#endif
+
+#include <stdlib.h>
+#include <math.h>
+
+#define XK_MISCELLANY
+#include <rfb/keysymdef.h>
+
+#include "GestureHandler.h"
+#include "BaseTouchHandler.h"
+
+// Sensitivity threshold for gestures
+static const int ZOOMSENS = 30;
+static const int SCRLSENS = 50;
+
+BaseTouchHandler::BaseTouchHandler()
+{
+}
+
+BaseTouchHandler::~BaseTouchHandler()
+{
+}
+
+void BaseTouchHandler::handleGestureEvent(const GHEvent& ev)
+{
+  double magnitude;
+
+  switch (ev.type) {
+  case GH_GestureBegin:
+    switch (ev.gesture) {
+    case GH_ONETAP:
+      fakeMotionEvent(ev);
+      fakeButtonEvent(true, 1, ev);
+      fakeButtonEvent(false, 1, ev);
+      break;
+    case GH_TWOTAP:
+      fakeMotionEvent(ev);
+      fakeButtonEvent(true, 3, ev);
+      fakeButtonEvent(false, 3, ev);
+      break;
+    case GH_THREETAP:
+      fakeMotionEvent(ev);
+      fakeButtonEvent(true, 2, ev);
+      fakeButtonEvent(false, 2, ev);
+      break;
+    case GH_DRAG:
+      fakeMotionEvent(ev);
+      fakeButtonEvent(true, 1, ev);
+      break;
+    case GH_LONGPRESS:
+      fakeMotionEvent(ev);
+      fakeButtonEvent(true, 3, ev);
+      break;
+    case GH_PAN:
+      lastMagnitudeX = ev.magnitude_x;
+      lastMagnitudeY = ev.magnitude_y;
+      fakeMotionEvent(ev);
+      break;
+    case GH_PINCH:
+      lastMagnitudeX = sqrt(ev.magnitude_x * ev.magnitude_x +
+                            ev.magnitude_y * ev.magnitude_y);
+      fakeMotionEvent(ev);
+      fakeKeyEvent(true, XK_Control_L, ev);
```

Comment:
I had to revert the commit about not keeping Ctrl pressed @CendioOssman. As I wrote in the commit msg:

> Windows does not handle keyboard and mouse events in order if sent from the same gesture event. All mouse events are handled before any keyboard events are handled.

Reply:
The problem was our AltGr handling on windows. Fixed now.